### PR TITLE
Improvement: Contributor nametag

### PR DESCRIPTION
**Pull request by ItsEmpa**
## What
Adds contributor suffix to nametags, and EntityDisplayNameEvent

<details>
<summary>Images</summary>

image
image

</details>

## Changelog Improvements
+ Added support for Fancy Contributors in nametags. - Empa

## Changelog Technical Details
+ Added EntityDisplayNameEvent. - Empa

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/misc/ContributorManager.kt [hunk 1]
```diff
@@ -16,7 +23,21 @@ object ContributorManager {
         contributors = event.getConstant<ContributorsJson>("Contributors").contributors.mapKeys { it.key.lowercase() }
     }
 
-    fun getTabListSuffix(username: String): String? = getContributor(username)?.suffix
+    @SubscribeEvent
+    fun onMobSpawnPlayer(event: MobEvent.Spawn.Player) {
```
Comment (ThatGravyBoat) [inline src/main/java/at/hannibal2/skyhanni/features/misc/ContributorManager.kt]: I dont think a spawn event should be used for such a cosmetic change, I think adding a new event for RenderEntityNameTag which allows overwriting the nametag would be better. This would also be more compatible with other mods

File: src/main/java/at/hannibal2/skyhanni/data/NametagData.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import at.hannibal2.skyhanni.events.EntityRenderNametagEvent
+import at.hannibal2.skyhanni.test.command.ErrorManager
+import net.minecraft.entity.Entity
+import net.minecraft.util.IChatComponent
+
+object NametagData {
+
+    fun getDisplayName(entity: Entity, name: IChatComponent): IChatComponent {
+        try {
+            if (EntityRenderNametagEvent(entity, name).postAndCatch()) {
```
Comment (ThatGravyBoat) [inline src/main/java/at/hannibal2/skyhanni/data/NametagData.kt]: `postAndCatch` already catches all exceptions. Remove this object and put the event directly in the mixin.

File: src/main/java/at/hannibal2/skyhanni/events/EntityRenderNametagEvent.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package at.hannibal2.skyhanni.events
+
+import net.minecraft.entity.Entity
+import net.minecraft.util.IChatComponent
+import net.minecraftforge.fml.common.eventhandler.Cancelable
+
+@Cancelable
+class EntityRenderNametagEvent(val entity: Entity, var chatComponent: IChatComponent) : LorenzEvent()
```
Comment (ThatGravyBoat) [inline src/main/java/at/hannibal2/skyhanni/events/EntityRenderNametagEvent.kt]: Cancelable is not needed as you never change operation depending on cancel.

The event should be named something along the lines of `EntityDisplayNameEvent`

The event should be put into the `entity` package of `events`

**Review by ThatGravyBoat — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/misc/ContributorManager.kt [hunk 4]
```diff
@@ -16,7 +19,15 @@ object ContributorManager {
         contributors = event.getConstant<ContributorsJson>("Contributors").contributors.mapKeys { it.key.lowercase() }
     }
 
-    fun getTabListSuffix(username: String): String? = getContributor(username)?.suffix
+    @SubscribeEvent
+    fun onRenderNametag(event: EntityRenderNametagEvent) {
+        if (event.entity !is EntityPlayer) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/ContributorManager.kt]: Also check for isRealPlayer here

**Review by ThatGravyBoat — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/events/entity/EntityDisplayNameEvent.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/events/entity/EntityDisplayNameEvent.kt:1]: ~~Not necessary, but I do like EntityRenderNametagEvent more than this~~ nvm

**Review by CalMWolfs — APPROVED** [review]

**Review by hannibal002 — DISMISSED** [review]
this needs a separate toggle (default off)

File: src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java [hunk 5]
```diff
@@ -63,12 +63,20 @@ public class DevConfig {
     @Expose
     @ConfigOption(
         name = "Fancy Contributors",
-        desc = "Marks §cSkyHanni's contributors §7fancy in the tab list. " +
+        desc = "Marks §cSkyHanni's contributors §7fancy in the tab list and nametags. " +
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java]:
```suggestion
        desc = "Marks §cSkyHanni's contributors §7fancy in the tab list. " +
```

File: src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java [hunk 6]
```diff
@@ -63,12 +63,20 @@ public class DevConfig {
...
     @ConfigEditorBoolean
     public boolean fancyContributors = true;
 
+    @Expose
+    @ConfigOption(
+        name = "Contributor Nametags",
+        desc = "Makes SkyHanni contributors' nametags fancy too. "
+    )
+    @ConfigEditorBoolean
+    public boolean contributorNametags = false;
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java]: Also @hannibal002 i think this should really be default on, not off

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java (hunk 6 above)
Comment (Nealkitt) [reply src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java]: I'm not currently fancy, but I feel like having this feature off by default is the same as not having it at all, the only people using it would be friends of contributors and contributors, the feature doesn't really make sense if it's off by default, plus the player already has the option of turning it off, this isn't like sbe where everyone can pay to be obnoxious, this is only for contributors so.
 tldr: if off, no one would turn it on except friends, if on, people who it annoys would turn it off

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java]: I agree with this argument.
@ItsEmpa feel free to turn it default on.

Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/config/features/dev/DevConfig.java]: alr, done

**Review by CalMWolfs — APPROVED** [review]
